cloud-controller-manager in crashloop due to nil pointer exception

cloud controller crashes in `github.com/equinix/cloud-provider-equinix-metal/metal.getMetalConfig` due to a nil pointer exception:
```
I0127 09:05:49.125494       1 serving.go:348] Generated self-signed cert in-memory
W0127 09:05:55.522563       1 authentication.go:423] failed to read in-cluster kubeconfig for delegated authentication: open /var/run/secrets/kubernetes.io/serviceaccount/token: no such file or directory
W0127 09:05:55.522602       1 authorization.go:226] failed to read in-cluster kubeconfig for delegated authorization: open /var/run/secrets/kubernetes.io/serviceaccount/token: no such file or directory
W0127 09:05:55.522616       1 authorization.go:194] No authorization-kubeconfig provided, so SubjectAccessReview of authorization tokens won't work.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x4b0cd4]

goroutine 1 [running]:
io.ReadAll({0x0, 0x0})
	/usr/local/go/src/io/io.go:661 +0xd4
io/ioutil.ReadAll(...)
	/usr/local/go/src/io/ioutil/ioutil.go:27
github.com/equinix/cloud-provider-equinix-metal/metal.getMetalConfig({_, _})
	/workspace/metal/config.go:100 +0xbd
github.com/equinix/cloud-provider-equinix-metal/metal.init.0.func1({0x0?, 0x0?})
	/workspace/metal/cloud.go:51 +0x5d
k8s.io/cloud-provider.GetCloudProvider({0x7ffecbf0a79e?, 0x8?}, {0x0, 0x0})
	/go/pkg/mod/k8s.io/cloud-provider@v0.25.0/plugins.go:86 +0xf9
k8s.io/cloud-provider.InitCloudProvider({0x7ffecbf0a79e, 0xc}, {0x0?, 0x8?})
	/go/pkg/mod/k8s.io/cloud-provider@v0.25.0/plugins.go:164 +0x2c5
main.cloudInitializer(0xc00011c820)
	/workspace/main.go:57 +0x4c
k8s.io/cloud-provider/app.NewCloudControllerManagerCommand.func1(0xc000004a00?, {0xc000145590?, 0x5?, 0x5?})
	/go/pkg/mod/k8s.io/cloud-provider@v0.25.0/app/controllermanager.go:88 +0x1e9
github.com/spf13/cobra.(*Command).execute(0xc000004a00, {0xc00011e130, 0x5, 0x5})
	/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:856 +0x67c
github.com/spf13/cobra.(*Command).ExecuteC(0xc000004a00)
	/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:974 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:902
main.main()
	/workspace/main.go:48 +0x2cc
```

Do I miss a critical config?

my deployment looks like:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: cloud-controller-manager
  name: cloud-controller-manager
spec:
  replicas: 1
  selector:
    matchLabels:
      app: cloud-controller-manager
  template:
    metadata:
      annotations:
        checksum/secret-cloudprovider: 90174e466bb73b0d9c3090e0fe4466ee0ba63cd4f2146562cabd422f96fde3f2
      labels:
        app: cloud-controller-manager
    spec:
      automountServiceAccountToken: false
      containers:
      - command:
        - ./cloud-provider-equinix-metal
        - --cloud-provider=equinixmetal
        - --leader-elect=false
        - --allow-untagged-cloud=true
        - --authentication-skip-lookup=true
        - --kubeconfig=/var/run/secrets/gardener.cloud/shoot/generic-kubeconfig/kubeconfig
        env:
        - name: METAL_API_KEY
          valueFrom:
            secretKeyRef:
              key: apiToken
              name: cloudprovider
        - name: METAL_PROJECT_ID
          valueFrom:
            secretKeyRef:
              key: projectID
              name: cloudprovider
        - name: METAL_METRO_NAME
          value: fr
        - name: METAL_LOAD_BALANCER
          value: metallb:///
        image: docker.io/equinix/cloud-provider-equinix-metal:v3.5.0
        imagePullPolicy: IfNotPresent
        name: cloud-provider-equinix-metal
        ports:
        - containerPort: 10253
          name: metrics
          protocol: TCP
        resources:
          limits:
            cpu: 500m
            memory: 512Mi
          requests:
            cpu: 100m
            memory: 64Mi
        volumeMounts:
        - mountPath: /var/run/secrets/gardener.cloud/shoot/generic-kubeconfig
          name: kubeconfig
          readOnly: true
      tolerations:
      - key: CriticalAddonsOnly
        operator: Exists
      volumes:
      - name: kubeconfig
        projected:
          defaultMode: 420
          sources:
          - secret:
              items:
              - key: kubeconfig
                path: kubeconfig
              name: generic-token-kubeconfig
              optional: false
          - secret:
              items:
              - key: token
                path: token
              name: shoot-access-cloud-controller-manager
              optional: false
```

--- a/metal/cloud.go
+++ b/metal/cloud.go
@@ -54,7 +54,7 @@ func init() {
 			return nil, fmt.Errorf("provider config error: %w", err)
 		}
 
-		// report the config
+		// report the config to startup logs
 		printMetalConfig(metalConfig)
 
 		// set up our client and create the cloud interface

--- a/metal/config.go
+++ b/metal/config.go
@@ -94,16 +94,21 @@ func override(options ...string) string {
 	return ""
 }
 
+// getMetalConfig returns a Config struct from a cloud config JSON file or environment variables
 func getMetalConfig(providerConfig io.Reader) (Config, error) {
 	// get our token and project
 	var config, rawConfig Config
-	configBytes, err := ioutil.ReadAll(providerConfig)
-	if err != nil {
-		return config, fmt.Errorf("failed to read configuration : %w", err)
-	}
-	err = json.Unmarshal(configBytes, &rawConfig)
-	if err != nil {
-		return config, fmt.Errorf("failed to process json of configuration file at path %s: %w", providerConfig, err)
+
+	// providerConfig may be nil if no --cloud-config is provided
+	if providerConfig != nil {
+		configBytes, err := ioutil.ReadAll(providerConfig)
+		if err != nil {
+			return config, fmt.Errorf("failed to read configuration : %w", err)
+		}
+		err = json.Unmarshal(configBytes, &rawConfig)
+		if err != nil {
+			return config, fmt.Errorf("failed to process json of configuration file at path %s: %w", providerConfig, err)
+		}
 	}
 
 	// read env vars; if not set, use rawConfig

--- a/metal/config_test.go
+++ b/metal/config_test.go
@@ -1,0 +1,128 @@
+package metal
+
+import (
+	"io"
+	"os"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+func Test_getMetalConfig(t *testing.T) {
+	type args struct {
+		providerConfig io.Reader
+	}
+
+	testKey, testProject := "test", "test"
+	defaultConfig := Config{
+		// not default, but required and set in tests
+		AuthToken: testKey,
+		ProjectID: testProject,
+		// defaults defined in getMetalConfig
+		LocalASN:                     DefaultLocalASN,
+		AnnotationLocalASN:           DefaultAnnotationNodeASN,
+		AnnotationPeerASN:            DefaultAnnotationPeerASN,
+		AnnotationPeerIP:             DefaultAnnotationPeerIP,
+		AnnotationSrcIP:              DefaultAnnotationSrcIP,
+		AnnotationBGPPass:            DefaultAnnotationBGPPass,
+		AnnotationNetworkIPv4Private: DefaultAnnotationNetworkIPv4Private,
+		AnnotationEIPMetro:           DefaultAnnotationEIPMetro,
+		AnnotationEIPFacility:        DefaultAnnotationEIPFacility,
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    Config
+		wantErr bool
+		env     map[string]string
+	}{
+		{
+			name: "no config",
+			args: args{
+				providerConfig: nil,
+			},
+			want:    defaultConfig,
+			wantErr: false,
+			env: map[string]string{
+				"METAL_API_KEY":    "test",
+				"METAL_PROJECT_ID": "test",
+			},
+		},
+		{
+			name: "empty config file",
+			args: args{
+				providerConfig: strings.NewReader(""),
+			},
+			want:    Config{},
+			wantErr: true,
+		},
+		{
+			name: "empty json config file",
+			args: args{
+				providerConfig: strings.NewReader("{}"),
+			},
+			want:    Config{},
+			wantErr: true, // environment variable "METAL_API_KEY" is required if not in config file
+		},
+		{
+			name: "empty json config file with env",
+			args: args{
+				providerConfig: strings.NewReader("{}"),
+			},
+			want:    defaultConfig,
+			wantErr: false,
+			env: map[string]string{
+				"METAL_API_KEY":    "test",
+				"METAL_PROJECT_ID": "test",
+			},
+		},
+		{
+			name: "json config",
+			args: args{
+				providerConfig: strings.NewReader(`{
+					"apiKey": "test",
+					"projectId": "test"
+				}`),
+			},
+			want:    defaultConfig,
+			wantErr: false,
+		},
+		{
+			name: "partial json config",
+			args: args{
+				providerConfig: strings.NewReader(`{
+					"apiKey": "test",
+				}`),
+			},
+			want:    Config{},
+			wantErr: true,
+		},
+		{
+			name: "invalid json",
+			args: args{
+				providerConfig: strings.NewReader(`{]`),
+			},
+			want:    Config{},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// TODO(displague) env is a global potentially polluting other tests. getMetalConfig should accept envs as an argument.
+			for k, v := range tt.env {
+				os.Setenv(k, v)
+			}
+			got, err := getMetalConfig(tt.args.providerConfig)
+			for k := range tt.env {
+				os.Unsetenv(k)
+			}
+			if (err != nil) != tt.wantErr {
+				t.Errorf("getMetalConfig() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("getMetalConfig() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}

